### PR TITLE
Rustdoc: Format constant values as rich-text and visibility-awarely

### LINKED ISSUE #98929: rustdoc: rich-text and visibility-aware formatting for constant values
While _types_ in rustdoc are richly formatted using HTML (most notably for hyperlinks to definitions),
_constant expressions_ and _values_ do not enjoy any such treatment and are printed as plain text (or
sometimes even entirely omitted from the documentation).
With the continuously growing language support for “`const`” (`const fn`, const generics, `adt_const_params`,
`generic_const_exprs`, `const_trait_impl`) which allows ever-more-complex constant expressions in
more and more places, it seems very fitting to improve rustdoc's output to help users navigate this complexity.

I propose the following:

* We start showing the constant value of free const items no matter their type. At present, only integers, bools and string literals are displayed.
* We try to use constant values as returned by `tcx.const_eval_poly` as much as possible for pretty-printing and only fall back on `hir::Expr` (unevaluated expressions) as a last resort. Currently for free and assoc. consts, rustdoc evaluates arbitrary const expressions of any type (arbitrary ADTs) but **throws away** the result if the type is not an integer (or similar) and instead prints the unevaluated expression.
* We try to evaluate const arguments and pretty-print the const value. Right now, we just print their unevaluated form, a `hir::Expr`.
* Regarding the formatting / pretty-printing of const values ([`ConstValue`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/interpret/enum.ConstValue.html) (or `ValTree`s now?)):
  * We now print struct and enum literals as one would expect. Examples: `S { f: 0 }`, `Ok(())`
  * However, we also **hide private and `doc(hidden)`** fields.
    In non-tuple structs, we skip the fields and add the symbol `..` (mirroring the syntax of struct patterns).
    In tuple structs, we replace them with `_` (just how it's already done today inside the code block at the top of the page of a tuple struct)
  * We make paths and struct fields **hyperlinks** to their definition.
  * We print a placeholder like `[/* N elements */]` for large arrays to not clutter the documentation
  * We print a placeholder like maybe `/* large string */` for large strings
  * ( Maybe we should also introduce a max-depth parameter similar to the one rustdoc has for types )

Unfortunately rustdoc's current reliance on `const_eval_poly` means that in a quite a lot of cases – namely those where the const expr contains type or const parameters (including the implicit type parameter `Self`) –, we won't get the benefits of this proposed richer pretty-printing since the evaluation is going to fail with the error _“too generic“_ and we need to fall back on printing the unevaluated `hir::Expr`. This is a known limitation of rustdoc and tracked in #82852.

Parts of this proposal aim to solve #97933 _properly_ whereas #98814 is but a hotfix.
I have already partially implemented this proposal in a yet to be published PR. You can get a visual taste of it below:

| Stable Rustdoc Output | Nightly Rustdoc Output (#98814) | Stalled WIP Implementation of this Proposal |
|---|---|---|
| ppconst_beta | ppconst_after_hotfix | ppconst_linkified  |

@rustbot label T-rustdoc A-rustdoc-ui C-enhancement
@rustbot claim

### LINKED ISSUE #99630: Rustdoc displays private internals of associated constants when reexporting types
`types/src/lib.rs`:
```rust
pub struct HasPrivateFields {
    _private: (),
}

impl HasPrivateFields {
    pub const ASSOC: Self = Self { _private: () };
}
```

`main/src/lib.rs`:
```rust
pub use types::HasPrivateFields;
```

This leaks private fields:
Doc page displaying private fields in associated constant initializers

Tested with `cargo clean && cargo +nightly doc`.

Note that this also applies to verbose initializers.

This issue was first brought up in #97933. The previous issue was closed due to a hotfix implemented in PR #98814. However, the hotfix fails to handle the case in this issue.

### Meta

`rustc +nightly --version --verbose`:
```
rustc 1.64.0-nightly (848090dcd 2022-07-22)
binary: rustc
commit-hash: 848090dcd18553b790461132ca9d2a020aeea9a2
commit-date: 2022-07-22
host: x86_64-pc-windows-msvc
release: 1.64.0-nightly
LLVM version: 14.0.6
```

### PR DESCRIPTION
**Pull request by fmease**
In rustdoc, we now attempt to display *arbitrary* constant values (e.g. structs, arrays). However, we do so *intelligently* by **not** showing private or `#[doc(hidden)]` fields and by creating hyperlinks to the definitions of structs, enums, enum variants, and struct & variant fields. Further, we won't show overly long arrays and (byte) string literals. Instead, we will display a placeholder.
As an aside, a constant _value_ is an _evaluated_ constant expression.

Fixes #98929.
Fixes #99630.

I expect that the UI changes will need some tweaking over time. Consider them an MVP.

The basic framework of the formatter is borrowed from the pretty-printers `pretty_print_const{,value}` in `rustc_middle`.
For now, we only pretty-print `rustc_middle::ty::Const`s and MIR `ConstValue`s. Valtrees are just displayed as `_` as I don't
think they surface in rustdoc at the moment since anonymous constants / constant arguments are not evaluated at the time of this writing (see #82852) and just stored as `hir::Expr`s.

@rustbot label T-rustdoc A-rustdoc-ui

CC @jsha
r? @GuillaumeGomez

---

[You can browse the generated docs of `src/test/rustdoc/const-value.rs` here](https://rustdoc.crud.net/fmease/rustdoc-fmt-const-vals-as-rich-text-vis-awarely/const-value/consts) (points of major interest: [`Struct`](https://rustdoc.crud.net/fmease/rustdoc-fmt-const-vals-as-rich-text-vis-awarely/const-value/consts/struct.Struct.html) and [`Protocol`](https://rustdoc.crud.net/fmease/rustdoc-fmt-const-vals-as-rich-text-vis-awarely/const-value/consts/trait.Protocol.html)).

### CODE REVIEW

**Comment by jyn514** [comment]
cc @rust-lang/wg-const-eval

**Comment by RalfJung** [comment]
> I pretty much copied & pasted pre-existing pretty-printing code and adapted it to the needs of rustdoc:

Is this necessary? The current const pretty-printing is not the result of a lot of careful design, it's just "good enough" for now. It is barely ever seen by users. Given that the rustdoc one will probably be polished, and is seen much more, I think it'd be great if the code could be shared, so that consts as printed in MIR can be (almost) as nice as what rustdoc does.

**Comment by fmease** [comment]
Thanks for giving some context. I get where you are coming from and I understand your concern about a potential lack of code sharing / potential code duplication.

I think, however, that rustdoc's version of the pretty-printer is going to heavily diverge from the original one both over the course of this PR and over the general lifetime of rustdoc. Already, there are some parts that drastically differ.

In the beginning, I did indeed copy & paste the printing logic but it was a mere starting point – and the only one I'd argue.
Now, my version does not concern itself in the slightest with faithfully rendering all different sorts of inference variables and placeholders, nor pointers and addresses. It simply outputs `_` instead. That'd be pretty useless as a debug format.

I really hope that some of my advancements will flow back into the original pretty-printer. Although most of the polishing you speak of will be very specific to rustdoc. Right now it consists of hiding private and hidden fields, hyperlink anchors and ellipses for long arrays and strings. I am not sure whether they fit your use case but I'd be happy to be corrected :)

File: src/librustdoc/clean/utils.rs
Comment (GuillaumeGomez) [inline src/librustdoc/clean/utils.rs:305]: If it's not used anymore, better remove it.

File: src/librustdoc/html/render/constant.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/constant.rs:87]: I strongly approve. Something like 2 or 3 as max depth?

File: src/librustdoc/html/render/context.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/context.rs:115]: `SharedContext` is accessible from `DocContext` which contains this information into its field `render_options`. Normally you shouldn't need to add these two fields here.

**Comment by GuillaumeGomez** [comment]
Overall it's really great! I didn't review too much into the details since we need to have everyone on board with this change but I'm really hopeful about this.

cc @rust-lang/rustdoc

**Comment by RalfJung** [comment]
@fmease that does sound reasonable; I wanted to make sure you thought about this. :)  Would be good to ensure the PR description reflects the final reaosning when the PR lands (as that will be embedded in the git history, too). I'm not familiar enough with the code to really make any concrete suggestions here. Let's see if @oli-obk has any ideas.

File: src/librustdoc/html/render/constant.rs
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/constant.rs:90]: It should indeed!

Comment (fmease) [reply src/librustdoc/html/render/constant.rs:90]: It should, yea. I have a FIXME for that already:

https://github.com/rust-lang/rust/blob/db16732cc1a6fa02779d2ff2e3e6d788bdacfc1b/src/librustdoc/html/render/constant.rs#L13-L14

**Comment by fmease** [comment]
(those were added bc of some rebase and and pre-push script issues)

@rustbot label -A-bootstrap -T-infra

File: src/rustdoc-json-types/lib.rs
Comment (aDotInTheVoid) [inline src/rustdoc-json-types/lib.rs:235]: What's the difference between `default` and `default_value`? Could you document it

File: src/librustdoc/json/conversions.rs
Comment (aDotInTheVoid) [inline src/librustdoc/json/conversions.rs:109]: I think it would be nicer to crate a new `Context` struct (like [SharedContext](https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/html/render/context/struct.SharedContext.html)) to put both the `Cache` and `tcx`, beacuse most of the time these are used their just calling more `from_tcx`, so I'd rather optimize to make calling quicer, and then the fields can be pulled out when needed. This also makes it easier if their's more shared state we need to add later.

File: src/librustdoc/json/conversions.rs
Comment (aDotInTheVoid) [inline src/librustdoc/json/conversions.rs:275]: If the depth/lenght is limited, that should be documented in `rustdoc-json-types`

File: src/rustdoc-json-types/lib.rs
Comment (fmease) [reply src/rustdoc-json-types/lib.rs:235]: `default_value` is the *evaluated* form of the default constant expression and `default` is an the *unevaluated* form.
Same story with the existing entries `value` and `expr` for free-standing constants.

Yeah, I should definitely document it if we want to keep it.
However, I'd rather get rid of `default_value` again and put the _evaluated_ expression into `default` entry. Technically also a breaking change I guess but since my hotfix PR #98814 unevaluated exprs have become quite useless (in most cases they are mapped to `_`).
What do you think?
Now that I think about it, #98814 also introduced a breaking change to the JSON output and that was gone unnoticed (no bump in the format version).

**Comment by oli-obk** [comment]
The only concern I have is that this could very well become a maintenance hazard. We'll now have to keep fixing the rustdoc const renderer along with the regular const pretty printer. We can see how well that goes in practice and revisit the question if it becomes a problem.

File: src/rustdoc-json-types/lib.rs
Comment (GuillaumeGomez) [reply src/rustdoc-json-types/lib.rs:235]: Then please comment it in the code. :)

**Comment by RalfJung** [comment]
> The only concern I have is that this could very well become a maintenance hazard. We'll now have to keep fixing the rustdoc const renderer along with the regular const pretty printer. We can see how well that goes in practice and revisit the question if it becomes a problem.

Hm, yeah that's a good point -- in particular given that this PR is *not* based on valtrees, which are intended to be the high-level representation we use for constants in the future. Ir is the intention that this can also print constants that cannot be converted into valtrees?

**Comment by oli-obk** [comment]
> is the intention that this can also print constants that cannot be converted into valtrees?

Yea, that's what I assumed. Tho it probably wouldn't be an issue if pointers can't get printed. The issue is that structs containing pointers don't get their other fields made available as a valtree

**Comment by fmease** [comment]
I am fully amenable to switching from MIR's `ConstValue` to Valtrees if it's possible and if it doesn't (drastically) decrease the number of different kinds of constants we can print in the current version of this PR. As previously mentioned, we don't want to print function and raw pointers anyways. I am not sure how much smaller the scope of Valtrees is compared to arbitrary constant values.

Historically, rustdoc has always used `const_eval_poly` which returns a `ConstValue`. That is probably the main reason why I didn't switch to Valtrees. If you want me to, I can try to switch to `eval_to_valtree`.

`const_eval_poly` is not a great choice for rustdoc anyway, since it doesn't handle polymorphic const expressions as you probably know. It really limits rustdoc's const support (#82852). Ideally, it would use `const_eval_resolve` and pass a `ParamEnv`.
However, rustdoc does not keep track of the latter at the moment.
Looking at `eval_to_valtree`, I notice that it unsurprisingly requires a `ParamEnv` (`ParamEnvAnd`), too. Not sure if I can just pass it an empty environment in my initial implementation without it leading to ICEs.

**Comment by RalfJung** [comment]
In some earlier sketches of valtrees and pattern compilation, we had thought of a "mixed" representation, where a valtree can be `Leaf` or `Branch` of `Opaque`, so that we can embed opaque bare constants in a valtree for those sub-trees that cannot be turned into valtrees themselves. This should probably be a generic type parameter of `Valtree` so that we can have valtrees where we know for sure there are no opaque parts in them. I think that is hat pattern matching ultimately wants, and that also seems like the perfect representation for rustdoc. I don't know if the current roadmap for const-eval still has anything like that -- @oli-obk and @lcnr have that roadmap in their head but I don't think it is written down anywhere. :)

@fmease long-term I don't think there is any alternative to proper ParamEnv tracking, in particular if you want to do anything with polymorphic consts.

**Comment by jyn514** [comment]
> However, rustdoc does not keep track of the latter at the moment.

That is not true. Rustdoc tracks param env here: https://github.com/rust-lang/rust/blob/e33cc71a61c91e1d510bf283e9d345067e64eed2/src/librustdoc/core.rs#L93-L102

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I think the current version is a great improvement. I expect some UI fixes to follow but we can see that later.

@RalfJung @oli-obk: Is it ok for you to approve this PR?

**Comment by RalfJung** [comment]
This is not changing the const-eval interpreter or its const representation in any way, and presumably uses the existing APIs in much the same way as the existing pretty-printer, so I don't think there is much I can contribute here in terms of reviewing -- unless you have questions about specific const-eval / interpreter APIs.

So, I leave the decision here up to the primary maintainers of the rustdoc crate.

@lcnr and @oli-obk know better than me what is planned for const representation and valtrees, which affect this much more than my own work on the interpreter internals.

**Comment by GuillaumeGomez** [comment]
Let's wait to see if one of the two has something to add then. :)

**Comment by lcnr** [comment]
my current goal would result in constants in patterns always being fully opaque using `PartialEq` and for us to not support partial valtrees. Not sure if that's possible, haven't had a lot of time for this recently.

This means that to print associated constants, you will have to continue using the ctfe representation as they might not be representable using valtrees. Constants in the type system, i.e. array lengths or arguments for const parameters, will all be representable using valtrees but the compiler has to be able to convert them back to the ctfe representation anyways as they might be used as values by const eval.

I think printing mir constants for rustdoc isn't great, but it seems unavoidable if you want to print associated consts. Also using that routine for type system consts is better than having two routines, one for valtrees and one for associated consts.

The approach therefore feels appropriate to me as long as the rustdoc team has to capacity to maintain this printing routine.

**Comment by GuillaumeGomez** [comment]
Ok, thanks for approving it. This is now something we need to discuss with @fmease and the other rustdoc team members then.

**Comment by GuillaumeGomez** [comment]
So we already have two approvals from here. The [discussion on zulip](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/new.20const.20display) seems stalled so do you think it's okay to merge this as a first step @rust-lang/rustdoc ?

**Comment by fmease** [comment]
This PR is currently blocked by an RFC I am in the process of writing as per Zulip discussion.

@rustbot blocked

**Comment by fmease** [comment]
I don't think it makes sense to keep this PR open, it's too bit-rotten. Further, this feature won't land as is anyway but hopefully as part of a more general and flexible mechanism dubbed `#[doc(const_display)]`. [See Zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/new.20const.20display/near/364308963).

**Pull request closed without merging**
